janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/GuavaSubqueryCache.java

```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.transaction.subquerycache;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.Weigher;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
+/**
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public class GuavaSubqueryCache implements SubqueryCache {
+    private Cache<JointIndexQuery.Subquery, SubqueryResult> indexCache;
+
+    public GuavaSubqueryCache(int concurrencyLevel, long maximumWeight) {
+        indexCache = CacheBuilder.newBuilder()
+            .weigher((Weigher<JointIndexQuery.Subquery, SubqueryResult>) (q, r) -> 2 + r.size())
+            .concurrencyLevel(concurrencyLevel).maximumWeight(maximumWeight).build();
+    }
+
+    @Override
+    public List<Object> getIfPresent(JointIndexQuery.Subquery key) {
+        int limit = key.getLimit();
+        JointIndexQuery.Subquery noLimitKey = key.updateLimit(0);
+        SubqueryResult result = indexCache.getIfPresent(noLimitKey);
+        if (result != null && (result.getLimit() >= limit || result.getLimit() > result.size())) {
+            return result.getValues().stream().limit(limit).collect(Collectors.toList());
```

Comment:
The Java Stream API ist relatively inefficient in most simple cases. Instead you could use `List.subList()` here.